From b204d76e9a64ec99f0151b7ca018f3b6df2bb1b7 Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Tue, 19 Sep 2017 11:30:50 +0300
Subject: [PATCH] sdk_dpaa: ls1043a errata: restrict the max mtu

We can not allow for Jumbo frames and large MTU values on LS1043A due to
the A-010022 FMan errata. All outgoing frames larger than 4K bytes are dropped.

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
---
 drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_common.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_common.c
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_common.c
@@ -263,7 +263,16 @@ EXPORT_SYMBOL(dpa_get_stats64);
 
 int dpa_change_mtu(struct net_device *net_dev, int new_mtu)
 {
-	const int max_mtu = dpa_get_max_mtu();
+	int max_mtu = dpa_get_max_mtu();
+
+#ifndef CONFIG_PPC
+	/* Due to the A010022 FMan errata, we can not use contig frames larger
+	 * than 4K, nor S/G frames. We need to prevent the user from setting a
+	 * large MTU.
+	 */
+	if (unlikely(dpaa_errata_a010022))
+		max_mtu = DPA_BP_RAW_SIZE;
+#endif
 
 	/* Make sure we don't exceed the Ethernet controller's MAXFRM */
 	if (new_mtu < 68 || new_mtu > max_mtu) {
